File: P090-iOSJS/OCJSMessageHandler/OCJSMessageHandler/WKWebViewController.m
//
//  WKWebViewController.m
//  OCJSMessageHandler
//
//  Created by andyron on 2017/9/6.
//  Copyright © 2017年 andyron. All rights reserved.
//

#import "WKWebViewController.h"
#import <WebKit/WebKit.h>
#import <AVFoundation/AVFoundation.h>

@interface WKWebViewController () <WKUIDelegate,WKScriptMessageHandler>

@property (strong, nonatomic) WKWebView *webView;

@end

@implementation WKWebViewController

- (void)viewDidLoad {
    [super viewDidLoad];

    [self initWKWebView];
}

- (void)initWKWebView
{
    WKWebViewConfiguration *configuration = [[WKWebViewConfiguration alloc] init];
    
    WKPreferences *preferences = [WKPreferences new];
    preferences.javaScriptCanOpenWindowsAutomatically = YES;
    preferences.minimumFontSize = 40.0;
    configuration.preferences = preferences;
    
    self.webView = [[WKWebView alloc] initWithFrame:self.view.frame configuration:configuration];
    
    
    NSString *urlStr = [[NSBundle mainBundle] pathForResource:@"index.html" ofType:nil];
    NSURL *fileURL = [NSURL fileURLWithPath:urlStr];
    [self.webView loadFileURL:fileURL allowingReadAccessToURL:fileURL];
    
    self.webView.UIDelegate = self;
    [self.view addSubview:self.webView];
}

- (void)viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];
    // addScriptMessageHandler 很容易导致循环引用
    // 控制器 强引用了WKWebView,WKWebView copy(强引用了）configuration， configuration copy （强引用了）userContentController
    [self.webView.configuration.userContentController addScriptMessageHandler:self name:@"Location"];
    [self.webView.configuration.userContentController addScriptMessageHandler:self name:@"Shake"];
    [self.webView.configuration.userContentController addScriptMessageHandler:self name:@"Pay"];
}

- (void)viewWillDisappear:(BOOL)animated
{
    [super viewWillDisappear:animated];
    
    [self.webView.configuration.userContentController removeScriptMessageHandlerForName:@"Location"];
    [self.webView.configuration.userContentController removeScriptMessageHandlerForName:@"Shake"];
    [self.webView.configuration.userContentController removeScriptMessageHandlerForName:@"Pay"];
}

#pragma mark - private method
- (void)getLocation
{
    // 获取位置信息
    
    // 将结果返回给js
    NSString *jsStr = [NSString stringWithFormat:@"setLocation('%@')",@"上海浦东新区"];
    [self.webView evaluateJavaScript:jsStr completionHandler:^(id _Nullable result, NSError * _Nullable error) {
        NSLog(@"%@----%@",result, error);
    }];
    
    NSString *jsStr2 = @"window.ctuapp_share_img";
    [self.webView evaluateJavaScript:jsStr2 completionHandler:^(id _Nullable result, NSError * _Nullable error) {
        NSLog(@"%@----%@",result, error);
    }];
}
- (void)shakeAction
{
    AudioServicesPlaySystemSound (kSystemSoundID_Vibrate);
}

- (void)payWithParams:(NSDictionary *)tempDic
{
    if (![tempDic isKindOfClass:[NSDictionary class]]) {
        return;
    }
    NSString *orderNo = [tempDic objectForKey:@"order_no"];
    long long amount = [[tempDic objectForKey:@"amount"] longLongValue];
    NSString *subject = [tempDic objectForKey:@"subject"];
    NSString *channel = [tempDic objectForKey:@"channel"];
    NSLog(@"%@---%lld---%@---%@",orderNo,amount,subject,channel);
    
    // 支付操作
    NSString *resStr = @"支付成功,order_no为";
    resStr = [resStr stringByAppendingString: orderNo];
    // 将支付结果返回给js
    NSString *jsStr = [NSString stringWithFormat:@"payResult('%@')",resStr];
    [self.webView evaluateJavaScript:jsStr completionHandler:^(id _Nullable result, NSError * _Nullable error) {
        NSLog(@"%@----%@",result, error);
    }];
}

#pragma mark - WKUIDelegate
- (void)webView:(WKWebView *)webView runJavaScriptAlertPanelWithMessage:(NSString *)message initiatedByFrame:(WKFrameInfo *)frame completionHandler:(void (^)(void))completionHandler
{
    
    UIAlertController *alert = [UIAlertController alertControllerWithTitle:@"提醒" message:message preferredStyle:UIAlertControllerStyleAlert];
    [alert addAction:[UIAlertAction actionWithTitle:@"知道了" style:UIAlertActionStyleCancel handler:^(UIAlertAction * _Nonnull action) {
        completionHandler();
    }]];
    
    [self presentViewController:alert animated:YES completion:nil];
}

#pragma mark - WKScriptMessageHandler
- (void)userContentController:(WKUserContentController *)userContentController didReceiveScriptMessage:(WKScriptMessage *)message
{
    //    message.body  --  Allowed types are NSNumber, NSString, NSDate, NSArray,NSDictionary, and NSNull.
    NSLog(@"body:%@",message.body);
    if ([message.name isEqualToString:@"Location"]) {
        [self getLocation];
    } else if ([message.name isEqualToString:@"Shake"]) {
        [self shakeAction];
    } else if ([message.name isEqualToString:@"Pay"]) {
        [self payWithParams:message.body];
    }
}
@end

File: 90-iOSJS/SwiftJSMessageHandler/SwiftJSMessageHandler/WKWebViewController.swift
//
//  WKWebViewController.swift
//  SwiftJSMessageHandler
//
//  Created by andyron on 2017/9/6.
//  Copyright © 2017年 andyron. All rights reserved.
//

import UIKit
import WebKit
import AVFoundation

class WKWebViewController: UIViewController, WKUIDelegate, WKScriptMessageHandler {
    
    lazy var webView: WKWebView = WKWebView(frame: self.view.frame)
    
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        initWKWebView()
    }
    
    func initWKWebView() {
        webView.configuration.preferences.javaScriptCanOpenWindowsAutomatically = true
        webView.configuration.preferences.minimumFontSize = 40.0
        
        let urlStr = Bundle.main.path(forResource: "index.html", ofType: nil)
        let fileURL = URL(fileURLWithPath: urlStr!)
        webView.loadFileURL(fileURL, allowingReadAccessTo: fileURL)
        
        webView.uiDelegate = self
        view.addSubview(webView)
    }

    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        
        webView.configuration.userContentController.add(self, name: "Location")
        webView.configuration.userContentController.add(self, name: "Shake")
        webView.configuration.userContentController.add(self, name: "Pay")
    }
    
    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)
        
        webView.configuration.userContentController.removeScriptMessageHandler(forName: "Location")
        webView.configuration.userContentController.removeScriptMessageHandler(forName: "Shake")
        webView.configuration.userContentController.removeScriptMessageHandler(forName: "Pay")
    }

    // MARK: -
    func getLocation() {
        let jsStr = "setLocation('上海浦东新区')"
        webView.evaluateJavaScript(jsStr) { (result, error) in
            print("\(String(describing: result))----\(error)")
        }
    }
    
    func shakeAction() {
        AudioServicesPlaySystemSound(kSystemSoundID_Vibrate)
    }
    
    func payWithParams(tempDic: Any) {
       
        guard tempDic is NSDictionary else {
            return
        }
        
        let jsStr = "payResult('支付成功')"
        webView.evaluateJavaScript(jsStr) { (result, error) in
            print("\(result)----\(error)")
        }
    }
    
    // MARK: - WKUIDelegate
    func webView(_ webView: WKWebView, runJavaScriptAlertPanelWithMessage message: String, initiatedByFrame frame: WKFrameInfo, completionHandler: @escaping () -> Void) {
        
        let alert = UIAlertController(title: "提醒", message: message, preferredStyle: .alert)
        alert.addAction(UIAlertAction(title: "知道了", style: .cancel, handler:{
            action in
            completionHandler()
        }))
        
        self.present(alert, animated: true, completion: nil)
    }
    
    // MARK: - WKScriptMessageHandler
    func userContentController(_ userContentController: WKUserContentController, didReceive message: WKScriptMessage) {
        
        if message.name == "Location" {
            getLocation()
        } else if message.name == "Shake" {
            shakeAction()
        } else if message.name == "Pay" {
            payWithParams(tempDic: message.body)
        }
    }
}
